lib/Sema/DerivedConformanceEquatableHashable.cpp

```diff
@@ -744,66 +745,159 @@ static Expr* integerLiteralExpr(ASTContext &C, int64_t value) {
-                                       combineResultExpr, /*implicit*/ true);
-  return assignExpr;
+  ASTContext &C = tc.Context;
+  auto parentDC = cast<DeclContext>(parentDecl);
+
+  // Expected type: (Self) -> (into: _UnsafeHasher) -> _UnsafeHasher
+  // Constructed as:
+  //   func type(input: Self,
+  //             output: func type(input: _UnsafeHasher,
+  //                               output: _UnsafeHasher))
+  // Created from the inside out:
+
+  Type hasherType = C.getUnsafeHasherDecl()->getDeclaredType();
+
+  // Params: self (implicit), hasher
+  auto *selfDecl = ParamDecl::createSelf(SourceLoc(), parentDC);
+  auto *hasherParam = new (C) ParamDecl(VarDecl::Specifier::Owned, SourceLoc(),
+                                        SourceLoc(), C.Id_into, SourceLoc(),
+                                        C.Id_hasher, hasherType, parentDC);
+  hasherParam->setInterfaceType(hasherType);
+
+  ParameterList *params[] = {ParameterList::createWithoutLoc(selfDecl),
+                             ParameterList::createWithoutLoc(hasherParam)};
+
+  // Func name: _hash(into: _UnsafeHasher)
+  DeclName name(C, C.Id_hash, params[1]);
+  auto *hashDecl = FuncDecl::create(C,
+                                    SourceLoc(), StaticSpellingKind::None,
+                                    SourceLoc(), name, SourceLoc(),
+                                    /*Throws=*/false, SourceLoc(),
+                                    nullptr, params,
+                                    TypeLoc::withoutLoc(hasherType),
+                                    parentDC);
+  hashDecl->setImplicit();
+  hashDecl->setBodySynthesizer(bodySynthesizer);
+
+  // Evaluate type of Self in (Self) -> (into: _UnsafeHasher) -> _UnsafeHasher
+  auto selfParam = computeSelfParam(hashDecl);
+  auto inputTypeElt = TupleTypeElt(hasherType, C.Id_into);
+  auto inputType = TupleType::get({inputTypeElt}, C);
+  auto innerType = FunctionType::get(inputType, hasherType,
+                                     FunctionType::ExtInfo());
+
+  Type interfaceType;
+  if (auto sig = parentDC->getGenericSignatureOfContext()) {
+    hashDecl->setGenericEnvironment(parentDC->getGenericEnvironmentOfContext());
+    interfaceType = GenericFunctionType::get(sig, {selfParam}, innerType,
+                                             FunctionType::ExtInfo());
+  } else {
+    // (Self) -> innerType == (_UnsafeHasher) -> _UnsafeHasher
+    interfaceType = FunctionType::get({selfParam}, innerType,
+                                      FunctionType::ExtInfo());
+  }
+  hashDecl->setInterfaceType(interfaceType);
+
+  // FIXME: We need to adjust the access level because of the backhanded way we
+  // synthesize _hash(into:) in deriveHashable. The access level wouldn't
+  // otherwise matter.
+
+  // hashDecl->copyFormalAccessAndVersionedAttrFrom(typeDecl);
+  hashDecl->setAccess(std::max(typeDecl->getFormalAccess(), AccessLevel::FilePrivate));
```

Comment:
This seems very fishy to me. Why can't we use `copyFormalAccessAndVersionedAttrFrom`? And can that go into the comment?

Reply:
Verily, it is extremely fishy. Adding some clarification:
```swift
 if (typeDecl->getFormalAccess() != AccessLevel::Private) {
    hashDecl->copyFormalAccessAndVersionedAttrFrom(typeDecl);
  } else {
    // FIXME: We want to call copyFormalAccessAndVersionedAttrFrom here, but we
    // can't copy the access level of a private type, because of the backhanded
    // way we synthesize _hash(into:) in deriveHashable -- we need to make sure
    // the resolver will find the new function, so it needs to be at least
    // fileprivate. (The access level of synthesized members doesn't normally
    // matter; they don't go through an access level check after being returned
    // from the synthesizer.)
    hashDecl->setAccess(std::max(typeDecl->getFormalAccess(), AccessLevel::FilePrivate));
    if (typeDecl->getAttrs().hasAttribute<VersionedAttr>()) {
      hashDecl->getAttrs().add(new (C) VersionedAttr(/*implicit=*/true));
    }
  }
```

Reply:
Ah, I'm pretty sure `copyFormalAccessAndVersionedAttrFrom` should just take a flag that says whether you're copying from the context or from a peer declaration. We've had this problem elsewhere.